### PR TITLE
Update how `tar` binaries are discovered

### LINKED ISSUE #648: Internal R tar is not handling long filenames.
Hi pacrat is using `tar="internal"`. It's perfect solution for compatibility between different systems but internal R tar is not handling long directory names.
https://github.com/rstudio/packrat/blob/3df1d62f77e98f846b11636b4a3203b925ec1028/R/restore.R#L481

Impact: RStudio products, Windows users

Example: Installing package from gitlab with long slug is not working. (https://gitlab.com/adam.forys/long-test-slug-skeleton)

Why it's happening? 

Gitlab endpoint `{gitlab_host}/api/v4/projects/{repository}/repository/archive?sha={sha}`
generate `tar.gz` file that contains directory created as `repository-slug-sha-sha/` so it gave `long-test-slug-skeleton-958296dbbbf7f1d82f7f5dd1b121c7558604809f-958296dbbbf7f1d82f7f5dd1b121c7558604809f` and `untar(tar="internal")` returns error `Error in file(name, "wb") : cannot open the connection`.

Installing package by packrat for repository slug that has more that 16 characters will fail.

Because packrat is soft deprecated it's only impact RStudio Connect/shinyapps etc. (Rstudio product). 

This problem is not an issue for `renv` as `renv` is detecting system tar. 

My recommendation is replace `"internal"` with `Sys.getenv("TAR", "internal")`.
```
untar(srczip, exdir = scratchDir, tar = Sys.getenv("TAR", "internal"))
```
After this change untar will be using system tar when set. 

Question: 
Is any chance this change will be apply in next Rstudio Connect release?

### PR DESCRIPTION
**Pull request by toph-allen**
This PR incorporates #678, but goes further, so will supersede that PR.

Fixes #648 

### Intent

Packrat has previously forced the use of R's internal tar implementation. This is buggy in a number of situations, including any repository with a slug > 16 characters.

Note: The code is ready for review, but QA notes are incomplete.

### Approach

Influenced by the approach taken in `renv`, we now define a function `tar_binary()` to encapsulate slightly more complex logic:

- If `TAR` is set, respect that;
- If on Unix, look for `tar` on the `PATH`;
- If on Windows, look for the system `tar` and use that if it exists (see [`renv` implementation](https://github.com/rstudio/renv/blob/2bea5cef7cb6c57e43580cf0a182cbe59ef3536a/R/tar.R#L14-L21))
- Otherwise, return `"internal"` and maybe emit a warning about long filenames.

I put the function in `env.R`, because it gathers information about the environment.

There were a number of locations where `tar()` or `untar()` were called with `internal`, and this PR changes all of them.

### Automated Tests

This PR adds automated tests. Note that it adds a dependency on `mockery` for the tests; I'm not sure if we're ok with this.

There are five test functions that test the logical flow of the function in the following cases:

- A `TAR` variable is defined returns the contents of that variable
- We're on Unix and we can find a `tar` binary on the `PATH` returns that binary
- We're on Unix and we can't find a `tar` binary on the `PATH` returns `"internal"`
- We're on Windows and we find system `tar` returns the system `tar`
- We're on Windows and we can't find system `tar` returns `"internal"`

### QA Notes

You can validate that `packrat::restore()` still works by installing the the version of the package from this branch and running `packrat::restore()` in an R session in the top level directory extracted from this .zip file: [simple-rmd.zip](https://github.com/rstudio/packrat/files/9253334/simple-rmd.zip)

This will perform a Packrat restore. I'm not sure if there's a way to validate that the download path has been exercised from `restore()` though.

You can also run `packrat:::tar_binary()` and see that it prints the expected path for your environment. For me, this looks like:

```r
> packrat:::tar_binary()
[1] "/usr/bin/tar"
```

### CODE REVIEW

File: R/env.R [hunk 1]
```diff
@@ -1,4 +1,38 @@
...
+      root <- "C:/Windows"
+    }
+    tar <- file.path(root, "System32/tar.exe")
+    if (file.exists(tar)) {
+      return(tar)
+    }
+  }
+
+  # Return internal only as a fallback with a warning.
+  warning("No external tar binary found. Using R's internal TAR, which may cause failures with long filenames.")
```
Comment (aronatkins) [inline R/env.R:31]: How frequently will this warning be emitted when using the `"internal"` tar?

**Review by aronatkins — APPROVED** [review]

**Review by aronatkins — COMMENTED** [review]
This needs NEWS.

**Comment by ChaitaC** [comment]
Packrat
```
RemoteType: github
RemoteHost: api.github.com
RemoteRepo: packrat
RemoteUsername: rstudio
RemoteRef: toph-untar-handling
RemoteSha: 108183392a8d733ed6bf66da65afab5b6de1c11c
GithubRepo: packrat
GithubUsername: rstudio
GithubRef: toph-untar-handling
GithubSHA1: 108183392a8d733ed6bf66da65afab5b6de1c11c
NeedsCompilation: no
Packaged: 2022-08-09 18:31:24 UTC; chaitamacpro
Built: R 4.1.1; ; 2022-08-09 18:31:25 UTC; unix

**Comment by ChaitaC** [comment]
verified on packrat branch`toph-untar-handling`

**Pull request merged**